File: packages/dara-core/cypress/dara/pages/trigger_variable.py
from dara.core.interactivity import DerivedVariable, Variable
from dara.components import Button, Card, Input, Stack, Text


def trigger_variable():
    """
    Test TriggerVariable action.
    """
    # Simple scenario - trigger a single DV with deps=[] - most common use case
    input_var = Variable(1)
    dv = DerivedVariable(lambda x: int(x) + 2, variables=[input_var], deps=[], uid='trigger_var_dv', cache=None)

    simple_scenario = Stack(
        Text('SIMPLE_INPUT:'),
        Input(value=input_var),
        Text('SIMPLE_OUTPUT:'),
        Text(dv),
        Button('SIMPLE_TRIGGER', onclick=dv.trigger()),
    )

    # Complex scenario - trigger a variable in a chain

    return Stack(Card(simple_scenario, title='Simple scenario'))

File: packages/dara-core/cypress/integration/trigger_variable.ts
import { loginBeforeRoute } from '../support/utils';

describe('Trigger Variable action', () => {
    beforeEach(loginBeforeRoute('/trigger_variable'));

    it('Simple TriggerVariable scenario', () => {
        // Because of debouncing we need to spy on the request and wait for it to fire before checking the output
        cy.intercept('/api/core/derived-variable/trigger_var_dv').as('trigger');

        cy.cardContent('Simple scenario').within(() => {
            cy.contains('div', 'SIMPLE_INPUT')
                .next()
                .find('input')
                .should('have.value', 1)
                .then(($input) => {
                    cy.contains('div', 'SIMPLE_OUTPUT').next().as('output').should('have.text', 3);

                    cy.wrap($input)
                        .clear()
                        .type('5')
                        .then(() => {
                            cy.get('@output').should('have.text', 3);
                        });

                    cy.wait(1000);

                    cy.contains('button', 'SIMPLE_TRIGGER')
                        .click()
                        .then(() => {
                            cy.get('@output').should('have.text', 7);
                        });
                });
        });
    });
});
